From a5bf3eedaf80714798a5d93ae031b995619bf351 Mon Sep 17 00:00:00 2001
From: Mike Looijmans <mike.looijmans@topic.nl>
Date: Tue, 17 Nov 2020 10:40:16 +0100
Subject: [PATCH] mmc: sdhci-of-arasan: Resolve failed sd/mmc tuning due to
 invalid code

Assigning set of default tapdelays with macros caused sd/emmc tuning
to fail. Implement properly with integer arrays.
---
 drivers/mmc/host/sdhci-of-arasan.c | 32 +++++++++++++++++-------------
 1 file changed, 18 insertions(+), 14 deletions(-)

diff --git a/drivers/mmc/host/sdhci-of-arasan.c b/drivers/mmc/host/sdhci-of-arasan.c
index 2d38a4325fff..ae9f9117d9ce 100644
--- a/drivers/mmc/host/sdhci-of-arasan.c
+++ b/drivers/mmc/host/sdhci-of-arasan.c
@@ -56,15 +56,15 @@
 #define SDHCI_ITAPDLY_ENABLE		0x100
 #define SDHCI_OTAPDLY_ENABLE		0x40
 
-#define ZYNQMP_ITAP_DELAYS {0x0, 0x15, 0x15, 0x0, 0x15, 0x0,\
-			    0x0, 0x3D, 0x12, 0x0, 0x0}
-#define ZYNQMP_OTAP_DELAYS {0x0, 0x5, 0x6, 0x0, 0x5, 0x3,\
-			    0x3, 0x4, 0x6, 0x3, 0x0}
+static const u32 ZYNQMP_ITAP_DELAYS[] = {0x0, 0x15, 0x15, 0x0, 0x15, 0x0,\
+					 0x0, 0x3D, 0x12, 0x0, 0x0};
+static const u32 ZYNQMP_OTAP_DELAYS[] = {0x0, 0x5, 0x6, 0x0, 0x5, 0x3,\
+					 0x3, 0x4, 0x6, 0x3, 0x0};
 
-#define VERSAL_ITAP_DELAYS {0x0, 0x2C, 0x2C, 0x0, 0x2C, 0x0,\
-			    0x0, 0x36, 0x1E, 0x0, 0x0}
-#define VERSAL_OTAP_DELAYS {0x0, 0x5, 0x4, 0x0, 0x4, 0x3,\
-			    0x2, 0x3, 0x5, 0x2, 0x0}
+static const u32 VERSAL_ITAP_DELAYS[] = {0x0, 0x2C, 0x2C, 0x0, 0x2C, 0x0,\
+					 0x0, 0x36, 0x1E, 0x0, 0x0};
+static const u32 VERSAL_OTAP_DELAYS[] = {0x0, 0x5, 0x4, 0x0, 0x4, 0x3,\
+					 0x2, 0x3, 0x5, 0x2, 0x0};
 
 #define MMC_BANK2		0x2
 
@@ -955,20 +955,24 @@ static void arasan_dt_parse_tap_delays(struct device *dev)
 	struct sdhci_host *host = platform_get_drvdata(pdev);
 	struct sdhci_pltfm_host *pltfm_host = sdhci_priv(host);
 	struct sdhci_arasan_data *sdhci_arasan = sdhci_pltfm_priv(pltfm_host);
-	u32 *itapdly;
-	u32 *otapdly;
+	u32 itapdly[MMC_TIMING_MMC_HS400 + 1];
+	u32 otapdly[MMC_TIMING_MMC_HS400 + 1];
 	int i;
 
 	if (of_device_is_compatible(pdev->dev.of_node, "xlnx,zynqmp-8.9a")) {
-		itapdly = (u32 [MMC_TIMING_MMC_HS400 + 1]) ZYNQMP_ITAP_DELAYS;
-		otapdly = (u32 [MMC_TIMING_MMC_HS400 + 1]) ZYNQMP_OTAP_DELAYS;
+		for (i = 0; i <= MMC_TIMING_MMC_HS400; i++)
+			itapdly[i] = ZYNQMP_ITAP_DELAYS[i];
+		for (i = 0; i <= MMC_TIMING_MMC_HS400; i++)
+			otapdly[i] = ZYNQMP_OTAP_DELAYS[i];
 		if (sdhci_arasan->mio_bank == MMC_BANK2) {
 			otapdly[MMC_TIMING_UHS_SDR104] = 0x2;
 			otapdly[MMC_TIMING_MMC_HS200] = 0x2;
 		}
 	} else {
-		itapdly = (u32 [MMC_TIMING_MMC_HS400 + 1]) VERSAL_ITAP_DELAYS;
-		otapdly = (u32 [MMC_TIMING_MMC_HS400 + 1]) VERSAL_OTAP_DELAYS;
+		for (i = 0; i <= MMC_TIMING_MMC_HS400; i++)
+			itapdly[i] = VERSAL_ITAP_DELAYS[i];
+		for (i = 0; i <= MMC_TIMING_MMC_HS400; i++)
+			otapdly[i] = VERSAL_OTAP_DELAYS[i];
 	}
 
 	arasan_dt_read_tap_delay(dev, itapdly, MMC_TIMING_SD_HS,
-- 
2.17.1

